checker/raw_result.go

```diff
@@ -97,12 +110,43 @@ type Issue struct {
 	// TODO: add fields, e.g., state=[opened|closed]
 }
 
+// Commit represents a commit.
+type Commit struct {
+	// Note: SHA is not directly accessible from a pull request.
+	// TODO:SHA          string
+	Committer User
+	// Note: not all commits are reviewed.
+	// `nil` indicate no reviews.
+	Review *Review
```

Comment:
I find it a bit non-intuitive to have `Review` inside `Commit`. A single PR can have multiple "commits" but the "review" is for a PR and not the commits individually.

I think it should be the other way around, i.e a `MergeCommit` (of type `Commit`) inside a `Review`. Wdut?

Reply:
Another suggestion here would be to update `CodeReviewData` to return both `[]Commits` and `[]PullRequests` (MergedPRs)

Reply:
re: counter-intuitive. A commit here is a commit to main branch. There's only one merge commit for a PR. So essentially, a commit to main may (or not) have an associated pull request.  Maybe I should rename to `MergeCommit` to be clearer? 

Is there a use case where we look a commits to non-main branch? Just thinking out lout whether the `Commit` field may be re-usable in the future. (we'll find out as we iterate :-))

re: to return both `[]Commits` and `[]PullRequests`. Do we gain something by doing this? I see a point for commits in general (if we have other use cases for it, maybe another check); but I don't immediately see that it helps for the Code-Review check. Can you clarify?

Thanks!

Reply:
Makes sense. In that case how about the below suggestion: 

```
Commit {
  MergeRequest *MergeRequest
  Committer *User
}

MergeRequest {
  Reviews []Review
  Author
  Number
}

Review {
  Author User
  Platform ReviewPlatform
  Status <APPROVED|DECLINED|...>
}
```

Here's my reasoning to suggest the update:

- There can be multiple reviews for a given commit which the current structure does not imply. 
- Every `Review` can have more info like `approved`, `declined` etc. which cannot be capture in the current structure.
- Reviews are associated with MergeRequest, not the other way round. i.e you can have a `MergeRequest` without a `Review`, but no `Review` without `MergeRequest`.

Let me know what you think.

Reply:
+1 on review status

Reply:
> Makes sense. In that case how about the below suggestion:

we can tweak it. Below a few more points

> ```
> Commit {
>   MergeRequest *MergeRequest
>   Committer *User
> }

a review is not always done thru a pull request, so MergeRequest must be inside a Review, not the other way around. Gerrit is an example: it reviews commit but does not use merge requests.

> MergeRequest {
>   Reviews []Review
>   Author
>   Number
> }
> 
> Review {
>   Author User
>   Platform ReviewPlatform
>   Status <APPROVED|DECLINED|...>
> }
> ```

I don't think we care about capturing multiple review systems for a single review. The project either uses GitHub reviews or Gerrit or Prow, we don't need Platform to be associated with each reviewer's review. 

> Here's my reasoning to suggest the update:
> 
> * There can be multiple reviews for a given commit which the current structure does not imply.

it is captured by the current design. A review can have multiple "authors". Authors can be changed to a more generic structure `ReviewData` that contains more than Authors, such as State or anything else we want to add in the future. 

> * Every `Review` can have more info like `approved`, `declined` etc. which cannot be capture in the current structure.

we can add a status, that's fair. I don't see a strong use case for it since the review system takes care of deciding when merging is acceptable, but I don't have a strong opinion on having it. It complicates the policy evaluation a little but not much. 

> * Reviews are associated with MergeRequest, not the other way round. i.e you can have a `MergeRequest` without a `Review`, but no `Review` without `MergeRequest`.

that's not the case, see comment above for Gerrit.

> Let me know what you think.

Reply:
let's finalize in our sync. Meanwhile, I'll add the State to the structure instead of just a list of Authors

Reply:
update: it's hard to have a review status for different review systems, because we detect the review system based on certain fields that indicate reviews. So we actually don't know what system a non-reviewed commit is using...
I suppose Gerrit and Prow have their ways of indicating a "need changes" or other status, but we don't support these atm. For context, to detect GitHub reviews we look at the merge request's `Status` field; to detect Gerrit reviews we look as labels' content; and so on.

1. I think most users will want to define policies based on what scorecard supports, like "use GitHub review" or "use GitHub review and the the bot named X is acceptable".

2. There is another category of policies for which scorecard has no support like "merge request has status X and comment message contains Y", which essentially tries to use the results to implement support for a review system we do not support. I am not against these uses cases, but I feel a review system should be detected within scorecard rather than being implemented as a policy. 

My original goal was to address case 1. If we don't make a judgement call on what a review means at all, this means we're aiming to address point 2 and in this case we cannot really expose a review field. Policy writers will need to write complicated policies (including us for the default one) that re-implement what a review means. I kinda feel this should be implemented in scorecard rather than a policy.

Maybe there's a way to address both by providing digested data for case 1; whilst still allowing case 2. I need to think a bit more about it.

Feel free to chime in :-)

Reply:
additional thoughts. Checks like Token-Permission, Dangerous-Workflows, etc may require the presence of case 2 policies, like "no secret in workflow unless action is one of X,Y,Z ". So maybe there's no way around supporting cases in category 2) above. This leads to other questions:
1. We'll have to build policy API/building blocks to help users write simple policies for case 1 policies. Essentially, this means writing all the check's evaluation using a policy language. 
2. The original idea was to be policy-language agnostic and leave it to users to write their policy with a policy language of their choice. That no longer seems to be the case if we have to support building blocks for developers to use.
3. What does a policy return? a Fail/Pass or a list of things that made the policy failed. I suppose the latter. What is the format of these results? Is it defined by the JSON input the policy is applied to, or can it be arbitrarily defined by us?
5. Our first customer is allstar and binary authorization. How will allstar evaluation a policy? Will it use its own policy language on top of the JSON raw output we provide? Does it expect scorecard to accept a policy file that scorecard evaluates itself (which goes against point 2 above)? 

Another approach we could consider is for the raw results to expose both complete raw data, and some data about what scorecard supports. In the context of this commit, it would be:
```
commits{ // raw results
  SHA
  MergeRequest{
   Number, Reviews{ Status, Authors}, MergeCommit, labels[]
  }
}

reviews[]{ // what scorecard supports
  review{
    commitSHA
    system (GitHub, Prow, etc)
    reviewers[]{}
  }
}
```

We could merge this PR this way to see where this takes us. I think we also we need a design doc.

Reply:
cc @asraa @jeffmendoza @oliverchang @chrismcgehee

Reply:
Setup a sync to discuss this over gvc.

Reply:
so I'll go ahead and do this:
```
commits{ 
  // Raw data
  SHA
  CommitMessage
  MergeRequest{
   Number, Reviews{ Status, Authors}, MergeCommit, labels[]
  }
  // "Opinionated" results
  approved-reviews{ // what scorecard supports
  system (GitHub, Prow, etc)
  []review{
    reviewers[]{ Login, }
  }
}
}
```

Reply:
> Setup a sync to discuss this over gvc.

We should push discussions publicly as much as possible.

Reply:
we've decided to use 
```
commits{ 
  SHA
  CommitMessage
  MergeRequest{
   Number, Reviews{ Status, Authors}, MergeCommit, labels[]
  }
```

and push the massaged version (easier to write policies for) in a different part of scorecard.